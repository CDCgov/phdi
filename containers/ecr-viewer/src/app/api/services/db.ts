import pgp from "pg-promise";

const db_url = process.env.DATABASE_URL || "";
export const database = pgp()(db_url);
