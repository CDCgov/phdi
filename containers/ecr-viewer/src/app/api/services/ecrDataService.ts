import pgPromise from "pg-promise";
import { GetObjectCommand, S3Client } from "@aws-sdk/client-s3";
import { streamToJson } from "@/app/api/services/utils";
import { database } from "@/app/api/services/db";

const S3_SOURCE = "s3";
const POSTGRES_SOURCE = "postgres";
const s3Client = new S3Client({ region: process.env.AWS_REGION });

/**
 * Get eCR data based on the ecrID.
 * @param ecrId - The id of the eCR.
 * @returns A promise resolving to eCR data.
 */
export const getEcrData = async (ecrId: string) => {
  if (process.env.SOURCE === S3_SOURCE) {
    return await getS3Data(ecrId);
  } else if (process.env.SOURCE === POSTGRES_SOURCE) {
    return await getPostgresData(ecrId);
  }
};

/**
 * Retrieves FHIR data from Postgres based on eCR ID.
 * @param ecrId - The id of the ecr.
 * @returns A promise resolving to eCR data.
 */
const getPostgresData = async (ecrId: string) => {
  const { ParameterizedQuery: PQ } = pgPromise;
  const findFhir = new PQ({
    text: "SELECT * FROM fhir WHERE ecr_id = $1",
    values: [ecrId],
  });
  return (await database.one(findFhir)).data;
};

/**
 * Retrieves FHIR data from S3 based on eCR ID.
 * @param ecrId - The id of the ecr.
 * @returns A promise resolving to eCR data.
 */
const getS3Data = async (ecrId: string) => {
  const bucketName = process.env.ECR_BUCKET_NAME;
  const objectKey = `${ecrId}.json`; // This could also come from the request, e.g., req.query.key
  const command = new GetObjectCommand({
    Bucket: bucketName,
    Key: objectKey,
  });

  const { Body } = await s3Client.send(command);
  return await streamToJson(Body);
};
