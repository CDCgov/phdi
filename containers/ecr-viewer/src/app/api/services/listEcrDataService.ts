import {
  S3Client,
  ListObjectsV2Command,
  ListObjectsV2CommandOutput,
} from "@aws-sdk/client-s3";
import { database } from "@/app/api/services/db";
import {
  formatDateTime,
  convertUTCToLocalString,
} from "@/app/services/formatService";

const S3_SOURCE = "s3";
const POSTGRES_SOURCE = "postgres";
const s3Client = new S3Client({ region: process.env.AWS_REGION });

export type ListEcr = {
  ecrId: string;
  dateModified: string;
}[];

/**
 * Handles GET requests by fetching data from different sources based on the environment configuration.
 * It supports fetching from S3 and Postgres. If the `SOURCE` environment variable is not set to
 * a supported source, it returns a JSON response indicating an invalid source.
 * @returns A promise that resolves to a `NextResponse` object
 *   if the source is invalid, or the result of fetching from the specified source.
 *   The specific return type (e.g., the type returned by `list_s3` or `list_postgres`)
 *   may vary based on the source and is thus marked as `unknown`.
 */
export async function listEcrData() {
  if (process.env.SOURCE === S3_SOURCE) {
    const data = await list_s3();
    return processListS3(data);
  } else if (process.env.SOURCE === POSTGRES_SOURCE) {
    const data = await list_postgres();
    console.log(data);
    return processListPostgres(data);
  } else {
    throw Error("Invalid Source");
  }
}

/**
 * Retrieves array of eCR IDs from PostgreSQL database.
 * @returns A promise resolving to a NextResponse object.
 */
const list_postgres = async () => {
  const listFhir =
    "SELECT fhir.ecr_id, date_created, patient_name_last, patient_name_last, patient_birth_date, report_date, reportable_condition FROM fhir LEFT OUTER JOIN fhir_metadata on fhir.ecr_id = fhir_metadata.ecr_id order by date_created DESC";
  try {
    return await database.manyOrNone(listFhir);
  } catch (error: any) {
    console.error("Error fetching data:", error);
    if (error.message == "No data returned from the query.") {
      throw Error("No eCRs found");
    } else {
      throw Error(error.message);
    }
  }
};

/**
 * Retrieves array of eCRs and their metadata from S3.
 * @returns A promise resolving to a NextResponse object.
 */
const list_s3 = async () => {
  const bucketName = process.env.ECR_BUCKET_NAME;

  try {
    const command = new ListObjectsV2Command({
      Bucket: bucketName,
    });

    return await s3Client.send(command);
  } catch (error: any) {
    console.error("S3 GetObject error:", error);
    throw Error(error.message);
  }
};

/**
 * Processes a list of eCR data retrieved from Postgres.
 * @param responseBody - The response body containing eCR data from Postgres.
 * @returns - The processed list of eCR IDs and dates.
 */
export const processListPostgres = (responseBody: any[]): ListEcr => {
  return responseBody.map((object) => {
    return {
      ecrId: object.ecr_id || "",
      dateModified: object.date_created
        ? convertUTCToLocalString(
            formatDateTime(new Date(object.date_created!).toISOString()),
          )
        : "",
    };
  });
};

/**
 * Processes a list of eCR data retrieved from an S3 bucket.
 * @param responseBody - The response body containing eCR data from S3.
 * @returns - The processed list of eCR IDs and dates.
 */
export const processListS3 = (
  responseBody: ListObjectsV2CommandOutput,
): ListEcr => {
  responseBody.Contents?.sort(
    (a, b) => Number(b.LastModified) - Number(a.LastModified),
  );

  return (
    responseBody.Contents?.map((object) => {
      return {
        ecrId: object.Key?.replace(".json", "") || "",
        dateModified: object.LastModified
          ? convertUTCToLocalString(
              formatDateTime(new Date(object.LastModified!).toISOString()),
            )
          : "",
      };
    }) || []
  );
};
