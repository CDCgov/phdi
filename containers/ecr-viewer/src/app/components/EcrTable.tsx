import React from "react";
import { listEcrData } from "@/app/api/services/listEcrDataService";

import { EcrTableClient } from "@/app/components/EcrTableClient";

const basePath =
  process.env.NODE_ENV === "production" ? process.env.NEXT_PUBLIC_BASEPATH : "";

/**
 * eCR Table
 * @param props - The properties passed to the component.
 * @param props.currentPage - The current page to be displayed
 * @param props.itemsPerPage - The number of items to be displayed in the table
 * @returns - eCR Table element
 */
const EcrTable = async ({
  currentPage,
  itemsPerPage,
}: {
  currentPage: number;
  itemsPerPage: number;
}) => {
  const startIndex = (currentPage - 1) * itemsPerPage;
  const initialData = await listEcrData(startIndex, itemsPerPage);
  const defaultSort = { columnId: "received-date", direction: "desc" };

  return (
    <div className="ecr-client-wrapper width-full overflow-auto">
      <EcrTableClient data={initialData} defaultSort={defaultSort} />
    </div>
  );
};

export default EcrTable;
