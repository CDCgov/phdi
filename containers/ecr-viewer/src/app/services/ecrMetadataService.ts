import {
  formatAddress,
  formatContactPoint,
  formatDateTime,
} from "@/app/services/formatService";
import { PathMappings, evaluateData } from "@/app/view-data/utils/utils";
import { Bundle, Organization } from "fhir/r4";
import { evaluate } from "@/app/view-data/utils/evaluate";
import { DisplayDataProps } from "@/app/view-data/components/DataDisplay";
import { evaluateReference } from "@/app/services/evaluateFhirDataService";

export interface ReportableConditions {
  [condition: string]: {
    [trigger: string]: Set<string>;
  };
}

export interface ERSDWarning {
  warning: string;
  versionUsed: string;
  expectedVersion: string;
  suggestedSolution: string;
}

/**
 * Evaluates eCR metadata from the FHIR bundle and formats it into structured data for display.
 * @param fhirBundle - The FHIR bundle containing eCR metadata.
 * @param mappings - The object containing the fhir paths.
 * @returns An object containing evaluated and formatted eCR metadata.
 */
export const evaluateEcrMetadata = (
  fhirBundle: Bundle,
  mappings: PathMappings,
) => {
  const rrDetails = evaluate(fhirBundle, mappings.rrDetails);

  let reportableConditionsList: ReportableConditions = {};

  for (const condition of rrDetails) {
    let name = condition.valueCodeableConcept.coding[0].display;
    const triggers = condition.extension
      .filter(
        (x: { url: string; valueString: string }) =>
          x.url ===
          "http://hl7.org/fhir/us/ecr/StructureDefinition/us-ph-determination-of-reportability-rule-extension",
      )
      .map((x: { url: string; valueString: string }) => x.valueString);
    if (!reportableConditionsList[name]) {
      reportableConditionsList[name] = {};
    }

    for (let i in triggers) {
      if (!reportableConditionsList[name][triggers[i]]) {
        reportableConditionsList[name][triggers[i]] = new Set();
      }
      condition.performer
        .map((x: { display: string }) => x.display)
        .forEach((x: string) =>
          reportableConditionsList[name][triggers[i]].add(x),
        );
    }
  }

  const custodianRef = evaluate(fhirBundle, mappings.eicrCustodianRef)[0] ?? "";
  const custodian = evaluateReference(
    fhirBundle,
    mappings,
    custodianRef,
  ) as Organization;

  const eicrReleaseVersion = (fhirBundle: any, mappings: any) => {
    const releaseVersion = evaluate(fhirBundle, mappings.eicrReleaseVersion)[0];
    console.log(releaseVersion);
    if (releaseVersion === "2016-12-01") {
      return "R1.1 (2016-12-01)";
    } else if (releaseVersion === "2021-01-01") {
      return "R3.1 (2021-01-01)";
    } else {
      return releaseVersion;
    }
  };

  const fhirERSDWarnings = evaluate(fhirBundle, mappings.eRSDwarnings);
  let eRSDTextList: ERSDWarning[] = [];

  for (const warning of fhirERSDWarnings) {
    if (warning.code === "RRVS34") {
      eRSDTextList.push({
        warning:
          "Sending organization is using an malformed eRSD (RCTC) version",
        versionUsed: "2020-06-23",
        expectedVersion:
          "Sending organization should be using one of the following: 2023-10-06, 1.2.2.0, 3.x.x.x.",
        suggestedSolution:
          "The trigger code version your organization is using could not be determined. The trigger codes may be out date. Please have your EHR administrators update the version format for complete eCR functioning.",
      });
    } else if (warning.code === "RRVS29") {
      eRSDTextList.push({
        warning:
          "Sending organization is using an outdated eRSD (RCTC) version",
        versionUsed: "2020-06-23",
        expectedVersion:
          "Sending organization should be using one of the following: 2023-10-06, 1.2.2.0, 3.x.x.x.",
        suggestedSolution:
          "The trigger code version your organization is using is out-of-date. Please have your EHR administration install the current version for complete eCR functioning.",
      });
    }
  }

  const eicrDetails: DisplayDataProps[] = [
    {
      title: "eICR ID",
      toolTip:
        "Unique document ID for the eICR that originates from the medical record. Different from the Document ID that NBS creates for all incoming records.",
      value: evaluate(fhirBundle, mappings.eicrIdentifier)[0],
    },
    {
      title: "Date/Time eCR Created",
      value: formatDateTime(
        evaluate(fhirBundle, mappings.dateTimeEcrCreated)[0],
      ),
    },
    {
      title: "eICR Release Version",
      value: eicrReleaseVersion(fhirBundle, mappings),
    },
    {
      title: "EHR Manufacturer Model Name",
      value: evaluate(fhirBundle, mappings.ehrManufacturerModel)[0],
    },
    {
      title: "EHR Software Name",
      toolTip: "EHR system used by the sending provider.",
      value: evaluate(fhirBundle, mappings.ehrSoftware)[0],
    },
  ];

  const ecrCustodianDetails: DisplayDataProps[] = [
    {
      title: "Custodian ID",
      value: custodian?.identifier?.[0]?.value,
    },
    {
      title: "Custodian Name",
      value: custodian?.name,
    },
    {
      title: "Custodian Address",
      value: formatAddress(
        custodian?.address?.[0].line ?? [],
        custodian?.address?.[0].city ?? "",
        custodian?.address?.[0].state ?? "",
        custodian?.address?.[0].postalCode ?? "",
        custodian?.address?.[0].country ?? "",
      ),
    },
    {
      title: "Custodian Contact",
      value: formatContactPoint(custodian?.telecom).join("\n"),
    },
  ];

  return {
    eicrDetails: evaluateData(eicrDetails),
    ecrCustodianDetails: evaluateData(ecrCustodianDetails),
    rrDetails: reportableConditionsList,
    eRSDWarnings: eRSDTextList,
  };
};
