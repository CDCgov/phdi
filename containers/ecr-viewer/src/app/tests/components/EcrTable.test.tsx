import { axe } from "jest-axe";
import { render } from "@testing-library/react";
import EcrTable from "@/app/components/EcrTable";
import { EcrDisplay, listEcrData } from "@/app/api/services/listEcrDataService";

jest.mock("../../api/services/listEcrDataService");

describe("EcrTable", () => {
  const mockedListEcrData = jest.mocked(listEcrData);
  const mockData: EcrDisplay[] = Array.from({ length: 51 }, (_, i) => ({
    ecrId: `id-${i + 1}`,
    patient_first_name: `first-${i + 1}`,
    patient_last_name: `last-${i + 1}`,
    dateModified: `2021-01-0${(i % 9) + 1}`,
    patient_date_of_birth: `2000-01-0${(i % 9) + 1}`,
    reportable_condition: `reportable-condition-${i + 1}`,
    rule_summary: `rule-summary-${i + 1}`,
    patient_report_date: `2021-01-0${(i % 9) + 1}`,
    date_created: `2021-01-0${(i % 9) + 1}`,
  }));

  beforeEach(() => {
    jest.resetAllMocks();
  });

  it("should match snapshot", async () => {
    mockedListEcrData.mockResolvedValue(mockData);
    const { container } = render(
      await EcrTable({ currentPage: 1, itemsPerPage: 25 }),
    );
    expect(container).toMatchSnapshot();
  });

  it("should pass accessibility", async () => {
    mockedListEcrData.mockResolvedValue(mockData);
    const { container } = render(
      await EcrTable({ currentPage: 1, itemsPerPage: 25 }),
    );
    expect(await axe(container)).toHaveNoViolations();
  });
});
