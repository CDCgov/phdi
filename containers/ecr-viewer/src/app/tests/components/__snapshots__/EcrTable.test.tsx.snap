// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EcrTable should match snapshot 1`] = `
<div>
  <div
    class="ecr-library-wrapper width-full overflow-auto"
  >
    <table
      class="usa-table usa-table--borderless width-full src-components-Table-Table-module__fixed--Eq00W usa-table--striped table-ecr-library margin-0"
      data-testid="table"
    >
      <thead
        class="position-sticky top-0"
      >
        <tr>
          <th
            class="minw-20"
            scope="col"
          >
            Patient
          </th>
          <th
            class="minw-1605"
            scope="col"
          >
            Received Date
          </th>
          <th
            class="minw-1705"
            scope="col"
          >
            Encounter Date
          </th>
          <th
            class="minw-2305"
            scope="col"
          >
            Reportable Condition
          </th>
          <th
            class="minw-23"
            scope="col"
          >
            RCKMS Rule Summary
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <a
              href="/view-data?id=id-1"
            >
              first-1
               
              last-1
            </a>
            <br />
            DOB: 2000-01-01
          </td>
          <td>
            2021-01-01
          </td>
          <td>
            2021-01-01
          </td>
          <td>
            reportable-condition-1
          </td>
          <td>
            rule-summary-1
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-2"
            >
              first-2
               
              last-2
            </a>
            <br />
            DOB: 2000-01-02
          </td>
          <td>
            2021-01-02
          </td>
          <td>
            2021-01-02
          </td>
          <td>
            reportable-condition-2
          </td>
          <td>
            rule-summary-2
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-3"
            >
              first-3
               
              last-3
            </a>
            <br />
            DOB: 2000-01-03
          </td>
          <td>
            2021-01-03
          </td>
          <td>
            2021-01-03
          </td>
          <td>
            reportable-condition-3
          </td>
          <td>
            rule-summary-3
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-4"
            >
              first-4
               
              last-4
            </a>
            <br />
            DOB: 2000-01-04
          </td>
          <td>
            2021-01-04
          </td>
          <td>
            2021-01-04
          </td>
          <td>
            reportable-condition-4
          </td>
          <td>
            rule-summary-4
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-5"
            >
              first-5
               
              last-5
            </a>
            <br />
            DOB: 2000-01-05
          </td>
          <td>
            2021-01-05
          </td>
          <td>
            2021-01-05
          </td>
          <td>
            reportable-condition-5
          </td>
          <td>
            rule-summary-5
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-6"
            >
              first-6
               
              last-6
            </a>
            <br />
            DOB: 2000-01-06
          </td>
          <td>
            2021-01-06
          </td>
          <td>
            2021-01-06
          </td>
          <td>
            reportable-condition-6
          </td>
          <td>
            rule-summary-6
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-7"
            >
              first-7
               
              last-7
            </a>
            <br />
            DOB: 2000-01-07
          </td>
          <td>
            2021-01-07
          </td>
          <td>
            2021-01-07
          </td>
          <td>
            reportable-condition-7
          </td>
          <td>
            rule-summary-7
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-8"
            >
              first-8
               
              last-8
            </a>
            <br />
            DOB: 2000-01-08
          </td>
          <td>
            2021-01-08
          </td>
          <td>
            2021-01-08
          </td>
          <td>
            reportable-condition-8
          </td>
          <td>
            rule-summary-8
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-9"
            >
              first-9
               
              last-9
            </a>
            <br />
            DOB: 2000-01-09
          </td>
          <td>
            2021-01-09
          </td>
          <td>
            2021-01-09
          </td>
          <td>
            reportable-condition-9
          </td>
          <td>
            rule-summary-9
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-10"
            >
              first-10
               
              last-10
            </a>
            <br />
            DOB: 2000-01-01
          </td>
          <td>
            2021-01-01
          </td>
          <td>
            2021-01-01
          </td>
          <td>
            reportable-condition-10
          </td>
          <td>
            rule-summary-10
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-11"
            >
              first-11
               
              last-11
            </a>
            <br />
            DOB: 2000-01-02
          </td>
          <td>
            2021-01-02
          </td>
          <td>
            2021-01-02
          </td>
          <td>
            reportable-condition-11
          </td>
          <td>
            rule-summary-11
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-12"
            >
              first-12
               
              last-12
            </a>
            <br />
            DOB: 2000-01-03
          </td>
          <td>
            2021-01-03
          </td>
          <td>
            2021-01-03
          </td>
          <td>
            reportable-condition-12
          </td>
          <td>
            rule-summary-12
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-13"
            >
              first-13
               
              last-13
            </a>
            <br />
            DOB: 2000-01-04
          </td>
          <td>
            2021-01-04
          </td>
          <td>
            2021-01-04
          </td>
          <td>
            reportable-condition-13
          </td>
          <td>
            rule-summary-13
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-14"
            >
              first-14
               
              last-14
            </a>
            <br />
            DOB: 2000-01-05
          </td>
          <td>
            2021-01-05
          </td>
          <td>
            2021-01-05
          </td>
          <td>
            reportable-condition-14
          </td>
          <td>
            rule-summary-14
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-15"
            >
              first-15
               
              last-15
            </a>
            <br />
            DOB: 2000-01-06
          </td>
          <td>
            2021-01-06
          </td>
          <td>
            2021-01-06
          </td>
          <td>
            reportable-condition-15
          </td>
          <td>
            rule-summary-15
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-16"
            >
              first-16
               
              last-16
            </a>
            <br />
            DOB: 2000-01-07
          </td>
          <td>
            2021-01-07
          </td>
          <td>
            2021-01-07
          </td>
          <td>
            reportable-condition-16
          </td>
          <td>
            rule-summary-16
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-17"
            >
              first-17
               
              last-17
            </a>
            <br />
            DOB: 2000-01-08
          </td>
          <td>
            2021-01-08
          </td>
          <td>
            2021-01-08
          </td>
          <td>
            reportable-condition-17
          </td>
          <td>
            rule-summary-17
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-18"
            >
              first-18
               
              last-18
            </a>
            <br />
            DOB: 2000-01-09
          </td>
          <td>
            2021-01-09
          </td>
          <td>
            2021-01-09
          </td>
          <td>
            reportable-condition-18
          </td>
          <td>
            rule-summary-18
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-19"
            >
              first-19
               
              last-19
            </a>
            <br />
            DOB: 2000-01-01
          </td>
          <td>
            2021-01-01
          </td>
          <td>
            2021-01-01
          </td>
          <td>
            reportable-condition-19
          </td>
          <td>
            rule-summary-19
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-20"
            >
              first-20
               
              last-20
            </a>
            <br />
            DOB: 2000-01-02
          </td>
          <td>
            2021-01-02
          </td>
          <td>
            2021-01-02
          </td>
          <td>
            reportable-condition-20
          </td>
          <td>
            rule-summary-20
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-21"
            >
              first-21
               
              last-21
            </a>
            <br />
            DOB: 2000-01-03
          </td>
          <td>
            2021-01-03
          </td>
          <td>
            2021-01-03
          </td>
          <td>
            reportable-condition-21
          </td>
          <td>
            rule-summary-21
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-22"
            >
              first-22
               
              last-22
            </a>
            <br />
            DOB: 2000-01-04
          </td>
          <td>
            2021-01-04
          </td>
          <td>
            2021-01-04
          </td>
          <td>
            reportable-condition-22
          </td>
          <td>
            rule-summary-22
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-23"
            >
              first-23
               
              last-23
            </a>
            <br />
            DOB: 2000-01-05
          </td>
          <td>
            2021-01-05
          </td>
          <td>
            2021-01-05
          </td>
          <td>
            reportable-condition-23
          </td>
          <td>
            rule-summary-23
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-24"
            >
              first-24
               
              last-24
            </a>
            <br />
            DOB: 2000-01-06
          </td>
          <td>
            2021-01-06
          </td>
          <td>
            2021-01-06
          </td>
          <td>
            reportable-condition-24
          </td>
          <td>
            rule-summary-24
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-25"
            >
              first-25
               
              last-25
            </a>
            <br />
            DOB: 2000-01-07
          </td>
          <td>
            2021-01-07
          </td>
          <td>
            2021-01-07
          </td>
          <td>
            reportable-condition-25
          </td>
          <td>
            rule-summary-25
          </td>
        </tr>
<<<<<<< HEAD
        <tr>
          <td>
            <a
              href="/view-data?id=id-26"
            >
              first-26
               
              last-26
            </a>
            <br />
            DOB: 2000-01-08
          </td>
          <td>
            2021-01-08
          </td>
          <td>
            2021-01-08
          </td>
          <td>
            reportable-condition-26
          </td>
          <td>
            rule-summary-26
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-27"
            >
              first-27
               
              last-27
            </a>
            <br />
            DOB: 2000-01-09
          </td>
          <td>
            2021-01-09
          </td>
          <td>
            2021-01-09
          </td>
          <td>
            reportable-condition-27
          </td>
          <td>
            rule-summary-27
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-28"
            >
              first-28
               
              last-28
            </a>
            <br />
            DOB: 2000-01-01
          </td>
          <td>
            2021-01-01
          </td>
          <td>
            2021-01-01
          </td>
          <td>
            reportable-condition-28
          </td>
          <td>
            rule-summary-28
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-29"
            >
              first-29
               
              last-29
            </a>
            <br />
            DOB: 2000-01-02
          </td>
          <td>
            2021-01-02
          </td>
          <td>
            2021-01-02
          </td>
          <td>
            reportable-condition-29
          </td>
          <td>
            rule-summary-29
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-30"
            >
              first-30
               
              last-30
            </a>
            <br />
            DOB: 2000-01-03
          </td>
          <td>
            2021-01-03
          </td>
          <td>
            2021-01-03
          </td>
          <td>
            reportable-condition-30
          </td>
          <td>
            rule-summary-30
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-31"
            >
              first-31
               
              last-31
            </a>
            <br />
            DOB: 2000-01-04
          </td>
          <td>
            2021-01-04
          </td>
          <td>
            2021-01-04
          </td>
          <td>
            reportable-condition-31
          </td>
          <td>
            rule-summary-31
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-32"
            >
              first-32
               
              last-32
            </a>
            <br />
            DOB: 2000-01-05
          </td>
          <td>
            2021-01-05
          </td>
          <td>
            2021-01-05
          </td>
          <td>
            reportable-condition-32
          </td>
          <td>
            rule-summary-32
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-33"
            >
              first-33
               
              last-33
            </a>
            <br />
            DOB: 2000-01-06
          </td>
          <td>
            2021-01-06
          </td>
          <td>
            2021-01-06
          </td>
          <td>
            reportable-condition-33
          </td>
          <td>
            rule-summary-33
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-34"
            >
              first-34
               
              last-34
            </a>
            <br />
            DOB: 2000-01-07
          </td>
          <td>
            2021-01-07
          </td>
          <td>
            2021-01-07
          </td>
          <td>
            reportable-condition-34
          </td>
          <td>
            rule-summary-34
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-35"
            >
              first-35
               
              last-35
            </a>
            <br />
            DOB: 2000-01-08
          </td>
          <td>
            2021-01-08
          </td>
          <td>
            2021-01-08
          </td>
          <td>
            reportable-condition-35
          </td>
          <td>
            rule-summary-35
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-36"
            >
              first-36
               
              last-36
            </a>
            <br />
            DOB: 2000-01-09
          </td>
          <td>
            2021-01-09
          </td>
          <td>
            2021-01-09
          </td>
          <td>
            reportable-condition-36
          </td>
          <td>
            rule-summary-36
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-37"
            >
              first-37
               
              last-37
            </a>
            <br />
            DOB: 2000-01-01
          </td>
          <td>
            2021-01-01
          </td>
          <td>
            2021-01-01
          </td>
          <td>
            reportable-condition-37
          </td>
          <td>
            rule-summary-37
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-38"
            >
              first-38
               
              last-38
            </a>
            <br />
            DOB: 2000-01-02
          </td>
          <td>
            2021-01-02
          </td>
          <td>
            2021-01-02
          </td>
          <td>
            reportable-condition-38
          </td>
          <td>
            rule-summary-38
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-39"
            >
              first-39
               
              last-39
            </a>
            <br />
            DOB: 2000-01-03
          </td>
          <td>
            2021-01-03
          </td>
          <td>
            2021-01-03
          </td>
          <td>
            reportable-condition-39
          </td>
          <td>
            rule-summary-39
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-40"
            >
              first-40
               
              last-40
            </a>
            <br />
            DOB: 2000-01-04
          </td>
          <td>
            2021-01-04
          </td>
          <td>
            2021-01-04
          </td>
          <td>
            reportable-condition-40
          </td>
          <td>
            rule-summary-40
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-41"
            >
              first-41
               
              last-41
            </a>
            <br />
            DOB: 2000-01-05
          </td>
          <td>
            2021-01-05
          </td>
          <td>
            2021-01-05
          </td>
          <td>
            reportable-condition-41
          </td>
          <td>
            rule-summary-41
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-42"
            >
              first-42
               
              last-42
            </a>
            <br />
            DOB: 2000-01-06
          </td>
          <td>
            2021-01-06
          </td>
          <td>
            2021-01-06
          </td>
          <td>
            reportable-condition-42
          </td>
          <td>
            rule-summary-42
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-43"
            >
              first-43
               
              last-43
            </a>
            <br />
            DOB: 2000-01-07
          </td>
          <td>
            2021-01-07
          </td>
          <td>
            2021-01-07
          </td>
          <td>
            reportable-condition-43
          </td>
          <td>
            rule-summary-43
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-44"
            >
              first-44
               
              last-44
            </a>
            <br />
            DOB: 2000-01-08
          </td>
          <td>
            2021-01-08
          </td>
          <td>
            2021-01-08
          </td>
          <td>
            reportable-condition-44
          </td>
          <td>
            rule-summary-44
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-45"
            >
              first-45
               
              last-45
            </a>
            <br />
            DOB: 2000-01-09
          </td>
          <td>
            2021-01-09
          </td>
          <td>
            2021-01-09
          </td>
          <td>
            reportable-condition-45
          </td>
          <td>
            rule-summary-45
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-46"
            >
              first-46
               
              last-46
            </a>
            <br />
            DOB: 2000-01-01
          </td>
          <td>
            2021-01-01
          </td>
          <td>
            2021-01-01
          </td>
          <td>
            reportable-condition-46
          </td>
          <td>
            rule-summary-46
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-47"
            >
              first-47
               
              last-47
            </a>
            <br />
            DOB: 2000-01-02
          </td>
          <td>
            2021-01-02
          </td>
          <td>
            2021-01-02
          </td>
          <td>
            reportable-condition-47
          </td>
          <td>
            rule-summary-47
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-48"
            >
              first-48
               
              last-48
            </a>
            <br />
            DOB: 2000-01-03
          </td>
          <td>
            2021-01-03
          </td>
          <td>
            2021-01-03
          </td>
          <td>
            reportable-condition-48
          </td>
          <td>
            rule-summary-48
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-49"
            >
              first-49
               
              last-49
            </a>
            <br />
            DOB: 2000-01-04
          </td>
          <td>
            2021-01-04
          </td>
          <td>
            2021-01-04
          </td>
          <td>
            reportable-condition-49
          </td>
          <td>
            rule-summary-49
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-50"
            >
              first-50
               
              last-50
            </a>
            <br />
            DOB: 2000-01-05
          </td>
          <td>
            2021-01-05
          </td>
          <td>
            2021-01-05
          </td>
          <td>
            reportable-condition-50
          </td>
          <td>
            rule-summary-50
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/view-data?id=id-51"
            >
              first-51
               
              last-51
            </a>
            <br />
            DOB: 2000-01-06
          </td>
          <td>
            2021-01-06
          </td>
          <td>
            2021-01-06
          </td>
          <td>
            reportable-condition-51
          </td>
          <td>
            rule-summary-51
          </td>
        </tr>
=======
>>>>>>> main
      </tbody>
    </table>
  </div>
</div>
`;
